{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "78732218",
   "metadata": {},
   "source": [
    "# Gene annotation\n",
    "#### Description\n",
    "This jupyter notebook contains a workflow to find the protein ID 'WP_011014180' of a specific gene (e.g. 'lysA')\n",
    "\n",
    "Input: \n",
    "- **genes_to_find** <class 'list'> to search for their corresponding protein IDs.\n",
    "- **model** <class 'cobra.core.model.Model'>\n",
    "\n",
    "Output:\n",
    "- **found_matches** <class 'dict'>\n",
    "\n",
    "\n",
    "#### Usage:\n",
    "\n",
    "```python\n",
    "genes_to_find = [\"ddh\", \"dapB\", \"lysA\"]\n",
    "found_matches = findGenes(genes_to_find, model)\n",
    "\n",
    "found_matches\n",
    "\n",
    "    {\n",
    "        'ddh': [],\n",
    "        'dapB': [{'id': 'WP_011014794', 'name': '4-hydroxy-tetrahydrodipicolinate reductase', 'similarity': None}],\n",
    "        'lysA': [{'id': 'WP_011014180', 'name': 'diaminopimelate decarboxylase', 'similarity': 100.0}], \n",
    "    }\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "#### Steps of the algorithm\n",
    "\n",
    "1. Building a local list **protein_id_to_gene_map** (.json file) containing all the published genes of **_C. glutamicum_ ATCC 13032** in RefSeq database\n",
    "2. Searching through protein_id_to_gene_map for matches based on **gene name**\n",
    "3. If genes still NOT found, collecting all the genes from homologous _C. glutamicum_ (10 strains available in KEGG)\n",
    "4. Searching through them for the gene name, when found, quering its amino acid sequence\n",
    "5. Searching through protein_id_to_gene_map for matches based on **AA sequence** homology\n",
    "\n",
    "Author: \n",
    "`Haroun Bensaadi`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b33db8b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from colorama import init, Fore, Style\n",
    "from collections import defaultdict\n",
    "from Bio import Entrez, SeqIO, pairwise2\n",
    "from IPython.display import Markdown\n",
    "import requests\n",
    "import time\n",
    "import json\n",
    "import re\n",
    "\n",
    "\n",
    "def findGenes(genes_to_find, model):\n",
    "    \n",
    "    \"\"\"\n",
    "    MAIN METHOD USED TO FIND THE GENES AND THEIR CORRESPONDING PROTEIN ID\n",
    "    \n",
    "    - INPUT: genes_to_find (type: list) \n",
    "    - OUTPUT: found_matches (type: json)\n",
    "\n",
    "    \"\"\"\n",
    "    global genes_not_found\n",
    "    global found_homologous_genes\n",
    "    \n",
    "    found_genes_by_gene_name = []\n",
    "    found_genes_by_aa_sequence = {}\n",
    "\n",
    "    found_matches = {gene: [] for gene in genes_to_find}\n",
    "    mapProteinIDtoGene()\n",
    "\n",
    "    # 1. searching by gene_name\n",
    "    for gene_name in genes_to_find:\n",
    "        match_results = searchGene(gene=gene_name)    \n",
    "        if match_results:\n",
    "            found_genes_by_gene_name.append(gene_name)\n",
    "            for match in match_results:\n",
    "                found_matches[gene_name].append({\"id\": match[0], \"name\": match[1][1], \"similarity\": None})\n",
    "                print(f\"    ● {match[0]} [{match[1][1]}]\")\n",
    "\n",
    "    # 2. searching by AA sequence homolgy\n",
    "    getDataBasedOnTaxonomy()\n",
    "    genes_not_found = list(set(genes_to_find) - set(found_genes_by_gene_name))\n",
    "\n",
    "    found_homologous_genes = getMatchedGenesByKEGGIdentifier()\n",
    "    aa_sequence_results = fetchAAsequenceOfHomologousGenes(found_homologous_genes)\n",
    "\n",
    "    print(\"\\nSearching for AA sequence homology against genes from C. glutamicum ATCC 13032:\")\n",
    "\n",
    "    for gene_id, data in list(aa_sequence_results.items()): \n",
    "        for element in data:\n",
    "            match_results = searchGene(seq= element[\"aa_seq\"], similarity_threshold=0.50, kegg_id = element[\"kegg_id\"])\n",
    "            if match_results:\n",
    "                for match in match_results:\n",
    "                    found_matches[gene_id].append({\"id\": match[0], \"name\": match[2][1], \"similarity\": float(match[1].split(\"similarity: \")[1][:-1])})\n",
    "                    print(f\"    ● {match[0]} ({match[1]}) [{match[2][1]}]: {match[2][3][:35]}...\")\n",
    "\n",
    "    # Keeping only the match with highest similarity score:\n",
    "    for gene, matches in found_matches.items():\n",
    "        if len(matches) > 1:\n",
    "            max_similarity = max(matches, key=lambda x: x['similarity'])\n",
    "            found_matches[gene] = [max_similarity]\n",
    "\n",
    "    return found_matches\n",
    "\n",
    "\n",
    "# 1. Direct search through the collected list of annoctated genes:¶\n",
    "\n",
    "def getFullGeneName(protein_id):\n",
    "    for gene in model.genes:\n",
    "        if protein_id in gene.id:\n",
    "            return gene.id\n",
    "        \n",
    "def getProteinID(full_gene):\n",
    "    return \"WP_\" + re.search(r\"WP_(\\d{9})\", full_gene).group(1) if re.search(r\"WP_(\\d{9})\", full_gene) else None\n",
    "\n",
    "def fetchProteinID(protein_id, max_retries=50, delay_between_retries=1):\n",
    "    for attempt in range(max_retries):\n",
    "        try:\n",
    "            Entrez.email = \"hte.bensaadi@gmail.com\"\n",
    "\n",
    "            handle = Entrez.efetch(db=\"protein\", id=protein_id, rettype=\"gb\", retmode=\"text\")\n",
    "            record = SeqIO.read(handle, \"genbank\")\n",
    "            handle.close()\n",
    "\n",
    "            gene_name, product, organism, sequence = None, None, None, None\n",
    "\n",
    "            for feature in record.features:\n",
    "                if 'gene' in feature.qualifiers:\n",
    "                    gene_name = feature.qualifiers['gene'][0]\n",
    "                if 'product' in feature.qualifiers:\n",
    "                    product = feature.qualifiers['product'][0]\n",
    "                if 'organism' in feature.qualifiers:\n",
    "                    organism = feature.qualifiers['organism'][0]\n",
    "\n",
    "            sequence = str(record.seq)\n",
    "\n",
    "            return protein_id, gene_name, product, organism, sequence\n",
    "\n",
    "        except Entrez.HTTPError as e:\n",
    "            if attempt < max_retries - 1:\n",
    "                print(f\"{protein_id}: retrying in {delay_between_retries} seconds...\")\n",
    "                time.sleep(delay_between_retries)\n",
    "            else:\n",
    "                print(f\"Max retries reached. Unable to fetch data for {protein_id}\")\n",
    "                raise\n",
    "        except Exception as e:\n",
    "            print(f\"Error: {e}\")\n",
    "            raise\n",
    "\n",
    "def loadProteinIDToGeneMap():\n",
    "    try:\n",
    "        with open(\"Files/protein_id_to_gene_map.json\", \"r\") as file:\n",
    "            protein_id_to_gene_map = json.load(file)\n",
    "    except FileNotFoundError:\n",
    "        protein_id_to_gene_map = {}\n",
    "    \n",
    "    return protein_id_to_gene_map\n",
    "    \n",
    "def updateProteinIDToGeneMap(protein_id_to_gene_map):\n",
    "    with open(\"Files/protein_id_to_gene_map.json\", \"w\") as file:\n",
    "        json.dump(protein_id_to_gene_map, file)\n",
    "           \n",
    "            \n",
    "def getNumberOfUniqueGenes():\n",
    "    \n",
    "    unique_protein_ids = []\n",
    "    \n",
    "    for gene in model.genes:\n",
    "        if getProteinID(gene.id) not in unique_protein_ids:\n",
    "            unique_protein_ids.append(getProteinID(gene.id))\n",
    "    return len(unique_protein_ids)\n",
    "    \n",
    "def mapProteinIDtoGene():\n",
    "    global protein_id_to_gene_map\n",
    "    \n",
    "    print(f\"Total number of unique genes in the model is: {getNumberOfUniqueGenes()} genes\")\n",
    "    \n",
    "    protein_id_to_gene_map = loadProteinIDToGeneMap()\n",
    "\n",
    "    if len(protein_id_to_gene_map) < 1:\n",
    "        print(f\"\\nFetching RefSeq data for C. glutamicum ATCC 13032: \")        \n",
    "\n",
    "        count_found_genes = 0\n",
    "        for gene in model.genes:\n",
    "            if \"WP_\" in gene.id:\n",
    "                protein_id = getProteinID(gene.id)\n",
    "\n",
    "                if protein_id not in protein_id_to_gene_map:\n",
    "                    protein_id, gene_name, product, organism, sequence = fetchProteinID(protein_id)\n",
    "                    protein_id_to_gene_map[protein_id] = [gene_name, product, organism, sequence]\n",
    "\n",
    "                    updateProteinIDToGeneMap(protein_id_to_gene_map)\n",
    "\n",
    "                    count_found_genes +=1\n",
    "\n",
    "                    print(f\"\\n{protein_id}, {gene_name}, {product}, {organism}, {sequence}\")\n",
    "\n",
    "        print(f\"Total number of fetched genes online is: {count_found_genes}\")\n",
    "    \n",
    "    else:\n",
    "        print(f\"\\nRefSeq data from  C. glutamicum ATCC 13032 is already stored locally: {len(protein_id_to_gene_map)+1} genes\")        \n",
    "\n",
    "# 2. Searching through AA sequence homology:\n",
    "\n",
    "def calculate_similarity(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2, one_alignment_only=True)\n",
    "    if alignments:\n",
    "        alignment = alignments[0]\n",
    "        alignment_length = max(len(alignment[0]), len(alignment[1]))\n",
    "        similarity = alignment[2] / alignment_length\n",
    "        return similarity\n",
    "    return 0\n",
    "\n",
    "def searchGene(gene=None, protein=None, seq=None, similarity_threshold=0.65, kegg_id=None):\n",
    "    global protein_id_to_gene_map\n",
    "    \n",
    "    matches = []\n",
    "    \n",
    "    for protein_id, data in protein_id_to_gene_map.items():\n",
    "        gene_name, product, organism, sequence = data\n",
    "        \n",
    "        if gene and gene_name and gene.lower() in gene_name.lower():\n",
    "            matches.append((protein_id, data))\n",
    "        \n",
    "        if protein and product and protein.lower() in product.lower():\n",
    "            matches.append((protein_id, data))\n",
    "        \n",
    "        if seq and sequence:\n",
    "            similarity = calculate_similarity(seq.lower(), sequence.lower())            \n",
    "            if similarity >= similarity_threshold:\n",
    "                matches.append((protein_id, f\"similarity: {round(similarity*100, 2)}%\", data))\n",
    "    \n",
    "    if gene and len(matches)>0:\n",
    "        print(f\"\\nMATCH: {gene} ->\\n\")\n",
    "        return matches\n",
    "\n",
    "    if seq and len(matches)>0:        \n",
    "        print(f\"\\nMATCH: {getGeneNameFromKEGGID(kegg_id)} ({kegg_id}) ->\\n\")\n",
    "        return matches\n",
    "\n",
    "    if protein and len(matches)>0:\n",
    "            print(f\"\\nMATCH: {protein} ->\\n\")\n",
    "            return matches\n",
    "\n",
    "    \n",
    "def getDataFromProtreinID(searched_protein_id):\n",
    "    protein_id_to_gene_map = loadMapFile()\n",
    "    for protein_id, data in protein_id_to_gene_map.items():\n",
    "        if searched_protein_id == protein_id:\n",
    "            return data\n",
    "\n",
    "def getGeneNameFromProteinID(searched_protein_id):\n",
    "    for gene, reactions in list(gene_to_reactions.items()):\n",
    "        if searched_protein_id == getProteinID(gene):\n",
    "            return gene\n",
    "    \n",
    "def getReactionsFromProteinID(searched_protein_id):\n",
    "    for gene, reactions in list(gene_to_reactions.items()):\n",
    "        if searched_protein_id == getProteinID(gene):\n",
    "            return reactions\n",
    "\n",
    "        \n",
    "def loadDataFromTaxonomySearch():\n",
    "    try:\n",
    "        with open(\"Files/data_from_taxonomy_search.json\", \"r\") as file:\n",
    "            data_from_taxonomy_search = json.load(file)\n",
    "    except FileNotFoundError:\n",
    "        data_from_taxonomy_search = []\n",
    "    return data_from_taxonomy_search\n",
    "    \n",
    "def saveDataFromTaxonomySearch(data_from_taxonomy_search):\n",
    "    with open(\"Files/data_from_taxonomy_search.json\", \"w\") as file:\n",
    "        json.dump(data_from_taxonomy_search, file)\n",
    "        \n",
    "def getDataBasedOnTaxonomy():\n",
    "    global data_from_taxonomy_search\n",
    "    taxonomy_codes = {\n",
    "        \"cgl\": \"Corynebacterium glutamicum ATCC 13032 (Kyowa Hakko)\",\n",
    "        \"cgb\": \"Corynebacterium glutamicum ATCC 13032 (Bielefeld)\",\n",
    "        \"cgu\": \"Corynebacterium glutamicum K051\",\n",
    "        \"cgt\": \"Corynebacterium glutamicum R\",\n",
    "        \"cgs\": \"Corynebacterium glutamicum SCgG1\",\n",
    "        \"cgg\": \"Corynebacterium glutamicum SCgG2\",\n",
    "        \"cgm\": \"Corynebacterium glutamicum MB001\",\n",
    "        \"cgj\": \"Corynebacterium glutamicum ATCC 21831\",\n",
    "        \"cgq\": \"Corynebacterium glutamicum AR1\",\n",
    "        \"cgx\": \"Corynebacterium glutamicum B253\"}\n",
    "\n",
    "    data_from_taxonomy_search = loadDataFromTaxonomySearch()\n",
    "    \n",
    "    if len(data_from_taxonomy_search) < 1:\n",
    "        print(f\"Fetching KEGG data for {len(taxonomy_codes)} C. glutamicum strain: \\n\")\n",
    "        \n",
    "        for taxonomy_code in taxonomy_codes:\n",
    "            search_url = f\"http://rest.kegg.jp/find/genes/{taxonomy_code}\"\n",
    "\n",
    "            try:\n",
    "                print(f\"Getting data for {taxonomy_codes[taxonomy_code]}\")\n",
    "                response = requests.get(search_url)\n",
    "                if response.status_code == 200:\n",
    "                    data_from_taxonomy_search.append(response.text.split('\\n'))\n",
    "                else:\n",
    "                    print(f\"Error: {response.status_code}\")\n",
    "            except requests.exceptions.RequestException as e:\n",
    "                print(f\"Request Exception: {e}\")\n",
    "\n",
    "        saveDataFromTaxonomySearch(data_from_taxonomy_search)\n",
    "    else:\n",
    "        print(f\"\\nKEGG data from other C. glutamicum strains is already stored locally: {len(data_from_taxonomy_search)} strains\\n\")\n",
    "            \n",
    "def getAminoAcidsSequence(gene_id):\n",
    "    response = requests.get(f\"http://rest.kegg.jp/get/{gene_id}\")\n",
    "    \n",
    "    if response.status_code == 200:        \n",
    "        lines = response.text.split('\\n')\n",
    "        amino_acid_sequence = \"\"\n",
    "        is_aa_sequence_section = False\n",
    "\n",
    "        for line in lines:\n",
    "            if line.startswith(\"AASEQ\"):\n",
    "                is_aa_sequence_section = True\n",
    "            elif line.startswith(\"NTSEQ\"):\n",
    "                is_aa_sequence_section = False\n",
    "            elif is_aa_sequence_section and line.strip(): \n",
    "                amino_acid_sequence += line.strip()\n",
    "        return amino_acid_sequence\n",
    "\n",
    "    else:\n",
    "        return f\"Error: {response.status_code}\"\n",
    "    \n",
    "    \n",
    "def getMatchedGenesByKEGGIdentifier():\n",
    "    global data_from_taxonomy_search\n",
    "    global genes_not_found\n",
    "    \n",
    "    found_homologous_genes = {}\n",
    "    \n",
    "    for gene_name_to_search in genes_not_found:\n",
    "        for data in data_from_taxonomy_search:\n",
    "            for entry in data:\n",
    "                \n",
    "                if re.compile(fr'{gene_name_to_search.lower()}.*?;').search(entry.lower()):\n",
    "                    if gene_name_to_search not in found_homologous_genes:\n",
    "                        found_homologous_genes[gene_name_to_search] = [entry]\n",
    "                    else:\n",
    "                        found_homologous_genes[gene_name_to_search].append(entry)\n",
    "\n",
    "        try:\n",
    "            print(f\"    Homologous genes for {gene_name_to_search}: {len(found_homologous_genes[gene_name_to_search])} results\")\n",
    "        except:\n",
    "            pass\n",
    "            \n",
    "    return found_homologous_genes\n",
    "\n",
    "\n",
    "def fetchAAsequenceOfHomologousGenes(found_homologous_genes):\n",
    "    \n",
    "    aa_sequence_results = {}\n",
    "    \n",
    "    print(\"\\nFetching AA sequence for the homologous genes: \\n\")\n",
    "\n",
    "    for gene_name, entries in list(found_homologous_genes.items()):\n",
    "        for entry in entries:\n",
    "            gene_id_kegg = entry.split(\"; \")[0].split(\"\\t\")[0]\n",
    "            gene_name_kegg = entry.split(\"; \")[0].split(\"\\t\")[1]\n",
    "            try:\n",
    "                gene_description_kegg = entry.split(\"; \")[1]\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "            aa_seq = getAminoAcidsSequence(gene_id_kegg)\n",
    "\n",
    "            print(f\"    {gene_name_kegg} ({gene_id_kegg}): {aa_seq[:25]}...\")\n",
    "\n",
    "            if gene_name not in aa_sequence_results:\n",
    "                aa_sequence_results[gene_name] = [{\"kegg_id\": gene_id_kegg, \"aa_seq\": aa_seq}]\n",
    "            else:\n",
    "                aa_sequence_results[gene_name].append({\"kegg_id\": gene_id_kegg, \"aa_seq\": aa_seq})\n",
    "\n",
    "    return aa_sequence_results\n",
    "\n",
    "def getGeneNameFromKEGGID(kegg_id_to_find):\n",
    "    for gene_name, entries in list(found_homologous_genes.items()):\n",
    "        for entry in entries:\n",
    "            kegg_id = entry.split(\"; \")[0].split(\"\\t\")[0]\n",
    "            gene_name = entry.split(\"; \")[0].split(\"\\t\")[1]\n",
    "            if kegg_id_to_find == kegg_id:\n",
    "                return gene_name\n",
    "\n",
    "\n",
    "def printReport(found_matches):    \n",
    "    GENES = {}\n",
    "    \n",
    "    print(\"\\n\\n\\nSummary:\\n\")\n",
    "    \n",
    "    for gene, matches in found_matches.items():\n",
    "        if len(matches) > 0 and matches[0]['similarity'] == None:\n",
    "            GENES[gene] = getFullGeneName(matches[0]['id'])\n",
    "            print(f\"    ● {gene}: {matches[0]['id']}\")\n",
    "    \n",
    "    for gene, matches in found_matches.items():    \n",
    "        if len(matches) > 0 and matches[0]['similarity']:\n",
    "            GENES[gene] = getFullGeneName(matches[0]['id'])\n",
    "            print(f\"    ● {gene}: {matches[0]['id']} (similarity {matches[0]['similarity']}%)\")\n",
    "    \n",
    "    for gene, matches in found_matches.items():\n",
    "        if len(matches) == 0:\n",
    "            print(\"\\nMissing genes:\\n\")    \n",
    "            print(f\"    ● {gene}: None\")\n",
    "\n",
    "    return GENES\n",
    "\n",
    "def printText(text, color=Fore.WHITE, style=Style.BRIGHT):\n",
    "    print(style + color + text + Style.RESET_ALL)\n",
    "    \n",
    "def printReaction(reaction):\n",
    "    substrates = ' + '.join([f\"{metabolite.name}\" for metabolite in reaction.reactants])\n",
    "    products = ' + '.join([f\"{metabolite.name}\" for metabolite in reaction.products])\n",
    "    reaction_string = f\"{substrates}  -->  {products}\"\n",
    "    return reaction_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19600f5d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
